### PR TITLE
Add rerank request deserialization

### PR DESCRIPTION
**Pull request by michalkulakowski**
### 🛠 Summary

https://jira.devtools.intel.com/browse/CVS-153700
https://jira.devtools.intel.com/browse/CVS-153705

### 🧪 Checklist

- [X] Unit tests added.
- [ ] The documentation updated.
- [ ] Change follows security best practices.
``

### CODE REVIEW

File: src/rerank/rerank.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#pragma once
+
+#include <limits>
```
Comment (dkalinowski) [inline src/rerank/rerank.hpp:18]: limits, set, sstream are not required here, rapidjson nethier, can be forward declared

File: src/rerank/rerank.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
+//*****************************************************************************
```
Comment (dkalinowski) [inline src/rerank/rerank.cpp:1]: rename to rerank_handler?

Comment (dkalinowski) [reply src/rerank/rerank.cpp:1]: or deserializer or utils

File: src/rerank/rerank.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        return absl::InvalidArgumentError("query is not a string");
+    } else {
+        request.query = it->value.GetString();
+    }
+    // documents: list or map of strings
+    it = doc.FindMember("documents");
+    if (it == doc.MemberEnd()) {
+        return absl::InvalidArgumentError("documents field is missing in request");
+    }
+    if (!it->value.IsArray()) {
```
Comment (dkalinowski) [inline src/rerank/rerank.cpp:54]: Comment states:
```
// documents: list or map of strings
```

But only list is implemented. Will we need both? @michalkulakowski @dtrawins

File: src/rerank/rerank.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            if (!d.IsString() && !d.IsObject()) {
+                return absl::InvalidArgumentError("documents array element is neither string nor object");
+            }
+            if (d.IsString()) {
+                if (request.documentsMap.size() > 0) {
+                    return absl::InvalidArgumentError("all documents have to be the same type (string or objects)");
+                }
+                request.documentsList.push_back(d.GetString());
+            }
+            if (d.IsObject()) {
```
Comment (dkalinowski) [inline src/rerank/rerank.cpp:67]:
```suggestion
            else if (d.IsObject()) {
```

File: src/rerank/rerank.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                request.documentsMap.insert({d.GetObject()["title"].GetString(), d.GetObject()["text"].GetString()});
+            }
+        }
+    }
+    // top_n: int; optional
+    it = doc.FindMember("top_n");
+    if (it != doc.MemberEnd()) {
+        if (!it->value.IsInt())
+            return absl::InvalidArgumentError("top_n accepts integer values");
+        request.topN = it->value.GetInt();
```
Comment (dkalinowski) [inline src/rerank/rerank.cpp:92]: Should we implement this on parser level?
```
defaults to the length of the documents
```

**Review by dkalinowski — APPROVED** [review]

**Review by rasapala — APPROVED** [review]

**Pull request merged**